Library/Homebrew/livecheck/options.rb

```diff
@@ -0,0 +1,105 @@
+      attr_reader :post_json
+
+      # @param homebrew_curl whether to use brewed curl
+      # @param post_form form data to use when making a `POST` request
+      # @param post_json JSON data to use when making a `POST` request
+      sig {
+        params(
+          homebrew_curl: T.nilable(T::Boolean),
+          post_form:     T.nilable(T::Hash[T.any(String, Symbol), String]),
+          post_json:     T.nilable(T::Hash[T.any(String, Symbol), String]),
+        ).void
+      }
+      def initialize(homebrew_curl: nil, post_form: nil, post_json: nil)
+        @homebrew_curl = homebrew_curl
+        @post_form = post_form
+        @post_json = post_json
+      end
+
+      # Returns a `Hash` of options that are provided as arguments to `url`.
+      sig { returns(T::Hash[Symbol, T.untyped]) }
+      def url_options
+        {
+          homebrew_curl: @homebrew_curl,
+          post_form:     @post_form,
+          post_json:     @post_json,
+        }
+      end
+
+      # Returns a `Hash` of all instance variable values.
+      sig { returns(T::Hash[Symbol, T.untyped]) }
+      def to_h
+        {
+          homebrew_curl: @homebrew_curl,
+          post_form:     @post_form,
+          post_json:     @post_json,
+        }
+      end
+
+      # Returns a `Hash` of all instance variables, using `String` keys.
+      sig { returns(T::Hash[String, T.untyped]) }
+      def to_hash
+        {
+          "homebrew_curl" => @homebrew_curl,
+          "post_form"     => @post_form,
+          "post_json"     => @post_json,
+        }
+      end
+
+      # Returns a new object formed by merging `other` values with a copy of
+      # `self`.
+      #
+      # `nil` values are removed from `other` before merging if it is an
+      # `Options` object, as these are unitiailized values. This ensures that
+      # existing values in `self` aren't unexpectedly overwritten with defaults.
+      sig { params(other: T.any(Options, T::Hash[Symbol, T.untyped])).returns(Options) }
+      def merge(other)
+        this_hash = to_h
+        other_hash = other.is_a?(Options) ? other.to_h.compact : other
+        return dup if this_hash == other_hash
+
+        new_options = this_hash.merge(other_hash)
+        Options.new(**new_options)
+      end
+
+      sig { params(other: Options).returns(T::Boolean) }
+      def ==(other)
+        instance_of?(other.class) &&
+          @homebrew_curl == other.homebrew_curl &&
+          @post_form == other.post_form &&
+          @post_json == other.post_json
+      end
+      alias eql? ==
+
+      # Whether the object has only default values.
+      sig { returns(T::Boolean) }
+      def empty?
+        @homebrew_curl.nil? && @post_form.nil? && @post_json.nil?
+      end
+
+      # Whether the object has any non-default values.
+      sig { returns(T::Boolean) }
+      def present?
+        !@homebrew_curl.nil? || !@post_form.nil? || !@post_json.nil?
+      end
```

Comment:
```suggestion
      def present? = !empty?
```

Reply:
I thought about implementing `present?` as the inverse of `empty?` but `empty?` has to check all values before returning. `present?` only needs to find one non-`nil` value, so it can potentially return early when it's implemented using "or" logic.

Reply:
This is a big reader tax, as maintainers need to mentally apply de morgan's laws to see the relationship between `present?` and `empty?`. For comparison, rails doesn't bother with these micro-optimizations, and just inverts related functions: https://github.com/rails/rails/blob/194d697036c61af0caa66de5659721ded2478ce9/activesupport/lib/active_support/core_ext/object/blank.rb#L18-L27